### ISSUE TITLE
Cannot bind SSDP port on OS X (UDP 1901) when occupied by Sonos Controller

### ISSUE THREAD

**Issue by vossavant** [jishi/node-sonos-http-api#53]
After running `npm install` and `node server.js` I see the following:
```
no settings file found, will only use default settings
error:  code=EADDRINUSE, errno=EADDRINUSE, syscall=bind
http server listening on port 5005
no preset file, ignoring...
```
I switch to my browser and enter in `http://localhost:5005/state` to get the state of the player, and then terminal outputs what you see here:

![sonos-http-error](https://cloud.githubusercontent.com/assets/1006875/7526305/add2f022-f4cd-11e4-8775-5e99e74413c1.png)

I guess this has something to do with how my SONOS player is configured. There are multiple players on the same network; I can access the web interface for my network at `http://10.0.1.51:1400/status`. I don't see anything in the docs on how to interface with my existing system, and I am pretty sure this is the missing link.

I am running node.js 0.12.2 on Max OS X Yosemite.

**Comment by jishi** [comment]
Hi. Seems like it doesn't find any players. The EADDRINUSE indicates that it has trouble binding some ports or similar, probably for discovery or for events. 

The port for discovery is UDP 1901, but usually this is not blocked. The port for events is 3500, and this is a TCP port and could collide with something else (maybe the official Sonos controller, although it should use 3400). I think you can use the following command to find which ports are listening and which process that started it:

sudo lsof -iTCP -sTCP:LISTEN -n -P

**Comment by jishi** [comment]
That verifies that port 3500 isn't occupied, which means that the cause is something else. I think you need to list UDP ports as well, so maybe replacing "TCP" to UDP in the command will give you the output for that. And also, if you could start the http-api and not invoke /state (to prevent the crash) and then give me the output of both TCP and UDP listening ports (with http-api running), it might tell me where the issue is.

**Comment by vossavant** [comment]
Hi,

Here is what I see after running `sudo lsof -iUDP -n -P` (I had to guess at the command - *no http-api running*):
![sonos-udp](https://cloud.githubusercontent.com/assets/1006875/7568258/f737c3e8-f7be-11e4-9674-6324ff9a46fd.png)

After starting the http-api and checking TCP, here is the output:
![sonos-api-running-tcp](https://cloud.githubusercontent.com/assets/1006875/7568373/a6e575d8-f7bf-11e4-841d-c96cadae9126.png)

And finally, after starting the http-api and checking UDP, here is the output:
![sonos-api-running-udp](https://cloud.githubusercontent.com/assets/1006875/7568405/df6428e6-f7bf-11e4-9fab-5ce00d2d2200.png)

Hopefully, that sheds some light.

**Comment by jishi** [comment]
Ah, yes it does. Seems like OS X doesn't allow multiple binds to UDP ports (opposed to Windows and Linux). I don't have an OS X machine at the moment, so I can't verify this. 

If you could test this: in file sonos.js (would be under node_modules/sonos-discovery/lib/sonos.js I think), could you change line 190 from:

var port = 1901;

to 

var port = 1902;

And start it again? And see if that resolves it? Then I can probably fix it so that it would find a free port automatically on OS X.

**Comment by vossavant** [comment]
Sweet! That worked. It's line 109, by the way, and not line 190.

Thank you for your help on this.

Do you by chance have any idea how I'd grab this information and send it up to a web server? The ultimate goal is to display the currently playing track on a website. I'm not looking for a detailed answer, just a rough idea if you have any.

You've been awesome - much appreciated!

**Comment by jishi** [comment]
Nice, thanks for testing it. 

Are you planning on posting it to a public web server? I have been thinking about adding an "url hook" feature to the http-api, where changes would be posted when something changes with the state or grouping, but haven't had time for it. 

You would probably be better of building it yourself around the sonos-discovery bundle, similar to this: https://github.com/jishi/sonos-event-stub

Depending on your js skills, this might be fairly trivial. In your case, you would only need to listen to the  'transport-state' event, which wuold include the same state-data that you receive from the http-api. It is evented on both the discovery object, and on individual players, so something like this:

**Comment by vossavant** [comment]
Yep, the idea is to post it to a public website. I have no idea how to get the local state of the player up to a remote server, and admittedly your stub doesn't answer that question for me, though I understand the `discovery.on('...')` syntax.

In other words, how would I listen to the state of the player (on my local network) from a remote server? I'll do some searching.

A URL hook would be great.

**Comment by jishi** [comment]
Well, you wouldn't listen from your remote server. The concept would be to have a local device that listen for state changes, and when identified, submit this data to your remote server. Meaning, running it on the same device that you run your http-api. Basically, you would build your own url-hook software.

**Comment by vossavant** [comment]
That makes sense. I've been trying to make sense of sonos-event-stub but can't figure out how to "compile" it - the "require" function throws an error (not defined), so I tried installing browserify, component, grunt... but nothing gets me a working script that I can link to in an HTML file to produce some kind of output.

In short, how do you go about actually using the event stub you created? When you wrote it, what did you use to run it? I understand this may require more hand holding than you are used to, so just tell me if so and I'll figure it out.

**Comment by jishi** [comment]
Ah, there is one step missing in the description. The package.json file includes dependencies that are required (sonos-discovery) and will be downloaded and installed automatically if you invoke:

npm install

in the same folder as the package.json file resides. This should work as long as the npm command is installed correctly (should be part of node.js). Then you would just start it by running:

node app.js

**Comment by vossavant** [comment]
Ah, thank you for that pointer! I'm not familiar with node and never would have guessed :)

So, after running `npm install` and `node app.js` I get the following:
![sonos-no-port-avail](https://cloud.githubusercontent.com/assets/1006875/7656161/0ee64422-fae6-11e4-92cd-51bb36263077.png)

I let it sit there for a long while, but nothing else happened. I didn't see anything in app.js that I could edit like we did before (changing port #). Any ideas?

**Comment by jishi** [comment]
Well, you have the same issue with this application on OS X. In the file node_modules/lib/sonos.js, line 109, you need to assigning a free UDP port to use for the discovery of the players.

Also note that this app doesn't do anything by itself, you will need to add the relevant actions on the transport-state event, like I previously stated. Something like:

**Comment by vossavant** [comment]
I've been playing around for a few days and I have things working. I'm getting the player state in Terminal and am working on sending the data via HTTP POST to a local script for processing.

The only trouble now is that when I `console.log` the state, so much information comes through that it's hard to parse the information I need. I.e.:

        discovery.on('transport-state', function (state) {
	        console.log(state);
        });

I'm really only interested in currentTrack...
![http-api-terminal-output](https://cloud.githubusercontent.com/assets/1006875/7731055/1b47283c-fedb-11e4-95af-49e2a20be7a6.png)

...But a whole bunch of other information from multiple rooms comes through. I think the overload of information is because we have 6 rooms in our office divided into 3 groups. Is there a way to just listen to one specific room?

In my `app.js` file, I have the `roomName` variable set to Development Den:
![http-api-code](https://cloud.githubusercontent.com/assets/1006875/7730941/71898a7e-feda-11e4-9a84-b6742d0d815f.png)

But in the output (<a href="http://3roadsmedia.com/http-api-console-output.txt">which you can see here as a TXT file</a>) there is track information from the other rooms, too, particuarly S&M.

Here is a screen shot from our SONOS app of our current configuration:
![sonos-groups](https://cloud.githubusercontent.com/assets/1006875/7730937/6c5963b2-feda-11e4-86db-eb34fdd3eba3.png)

I'd like to just get the state of *Development Den* when tracks change; right now I think I get a state update whenever a room in any group moves to the next track.

**Comment by jishi** [comment]
Yes, you are correct. Since you are subscribing to the discovery-object directly, you receive updates for all players in your Sonos system. You can subscribe to only a specific player, but I'm not sure how it behaves when players are grouped. I know that the coordinator will send updates, but I'm not sure that the other grouped players do that as well. However, I do think that would be preferable so I might as well fix that.

To subscribe to only a specific player, requires that you wait until you have discovered the players, and then get the player by name, and subscribe to the 'transport-state' event, like this:

    discovery.on('topology-change', function () {
        var player = discovery.getPlayer('Development Den');
        player.on('transport-state', function (state) {
            // Do stuff here
        });
    });

**Comment by vossavant** [comment]
Thanks for all of your help. I ended up using the following code to listen on transportChange:

        function transportChangeHandler(transportData) {
            var state = transportData.state,
                  requestify = require('requestify');

            requestify.post('http://www.example.com/sonos_webhook.php', {
                data: transportData,
                formName: 'GSCsonosPostData'
            })
            .then(function(response) {
                var phpReponse = JSON.parse(response.body);
                console.log(response.body);
            });
        });

I am using Requestify to make the HTTP POST.

Then, in my PHP (edited for clarity):

        <?php
        $sonos_data = json_decode( file_get_contents( 'php://input' ) );
        $room_name = $sonos_data->data->roomName;
        
        if ( $room_name == "Development Den" ) :
            $current_track = $sonos_data->data->state->currentTrack;
            ... save track info to database ...

**Comment by jishi** [comment]
Switched to port 1905 to not conflict with Sonos app
